src/term/string.rs

```diff
@@ -0,0 +1,399 @@
+            });
+        }
+
+        result.into()
+    }
+
+    /// Returns the substring of `self` between `start` and `end`.
+    ///
+    /// Returns an error if:
+    ///    - either start or end is not a fraction
+    ///    - start < 0
+    ///    - start > end
+    ///
+    /// The time complexity of this method is `O(self.len())`.
+    pub fn substring(&self, start: &Rational, end: &Rational) -> Result<NclString, SubstringError> {
+        let Ok(start_usize) = usize::try_from(start) else {
+            return Err(SubstringError::NonIntStart(start.clone()));
+        };
+
+        let Ok(end_usize) = usize::try_from(end) else {
+            return Err(SubstringError::NonIntEnd(end.clone()));
+        };
+
+        // We need to know the length of the grapheme cluster iterator,
+        // and we have to walk the graphemes to do that, so we might
+        // as well build a Vec to make slicing out the substring easier.
+        let graphemes: Vec<_> = self.graphemes(true).collect();
+        let str_len = graphemes.len();
+
+        if start < &0 || start_usize > str_len {
+            Err(SubstringError::StartOutOfBounds {
+                start: start.clone(),
+                str_len,
+            })
+        } else if end < start || end_usize > str_len {
+            Err(SubstringError::EndOutOfBounds {
+                start: start.clone(),
+                end: end.clone(),
+                str_len,
+            })
+        } else {
+            Ok(graphemes[start_usize..end_usize].concat().into())
+        }
+    }
+
+    /// Consumes `self`, returning the Rust `String`.
+    pub fn into_inner(self) -> String {
+        self.0
+    }
+}
+
+/// Errors returned by `NclString`'s `substring` method.
+pub enum SubstringError {
+    /// The start index was not an int
+    NonIntStart(Rational),
+    /// The end index was not an int
+    NonIntEnd(Rational),
+    /// The start index was not within the bounds of the string
+    StartOutOfBounds { start: Rational, str_len: usize },
+    EndOutOfBounds {
+        start: Rational,
+        end: Rational,
+        str_len: usize,
+    },
+}
+
+impl std::fmt::Display for SubstringError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use SubstringError::*;
+
+        write!(f, "substring: ")?;
+
+        match self {
+            NonIntStart(start) => write!(
+                f,
+                "expected 2nd argument (start) to be a positive integer smaller than {}, got {start}",
+                usize::MAX
+            ),
+            NonIntEnd(end) => write!(
+                f,
+                "expected 3rd argument (end) to be a positive integer smaller than {}, got {end}",
+                usize::MAX
+            ),
+            StartOutOfBounds { start, str_len } => write!(
+                f,
+                "index out of bounds. 
+                 Expected 2nd argument (start) to be between 0 and {str_len}, got {start}"
```

Comment:
I fear indentation is included in the string, which would give strange result on the terminal (unfortunately Rust doesn't have Nickel's multiline strings :p). If you want to split things a bit, maybe `{}\n{}` followed by your two lines? Or your can also use a trailing `\` as well. This will strip all the indentation, so it's not as smart as Nickel, but as long you only use one indentation level for the whole string, this should be fine:

```suggestion
                "index out of bounds.\n\
                 Expected 2nd argument (start) to be between 0 and {str_len}, got {start}"
```

Reply:
I actually meant to put these on the same line but just split up the text so it's easier to read. I just forgot to escape the newline. :sweat_smile: